File: CSE12/hw6/Driver.c
/**************************************************************************** 
 
                                                                Zifan Zhang 
                                                         	CSE 12, Fall 19 
                                                                11/5/2019 
                                                             	cs12fa19gt 
                		Assignment SIX
 
File Name:  	Driver.c
Description:    This pogram implement UCSDStudent to store student numbers
		with the associated student name
****************************************************************************/
#include <string.h>
#include <stdio.h>
#include <iostream>
#include <getopt.h>
#include "Base.h"
#include "SymTab.h"
#include <cstdlib>
using namespace std;

#ifdef NULL
#undef NULL
#define NULL 0
#endif
//=========================================================================
// class UCSDStudent : public Base
//
// Description: Implements the UCSDStudent. Each UCSDStudent has a name
// 		and a student number. 
//
// Data Fields:
//     name (char *)  - name of the UCSD student
//     studentnum (long)  - the student number of the UCSD student
//
// Public functions:
//     UCSDSudent - allocae memory and initialize the UCSDStudent
//     ~UCSDStudent - delete memory of the UCSDStudent
//     operator const char * - override char * casting of UCSDStudent
//     				return its name
//     operator long - override long casting of UCSDStudent return its hase
//     			code
//     operator == - override == operator return if two Bases's name are equal
//     operator < - override < operator return if one Base is less than another
//     
//==========================================================================
class UCSDStudent : public Base {
	char * name;
	long studentnum;
public:
/***************************************************************************
% Routine Name : UCSDStudent (char * nm, long sn)
% File :         Driver.c
% 
% Description :  This function allocate and initial memory for UCSDStudent
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
% nm	          	name of the UCSDStudent
% sn			student number of the UCSDStudent
***************************************************************************/
	UCSDStudent (char * nm, long sn) :
		name (strdup (nm)), studentnum (sn) {}
/***************************************************************************
% Routine Name : ~UCSDStudent (void)
% File :         Driver.c
% 
% Description :  This function delete memory of UCSDStudent
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
***************************************************************************/
	~UCSDStudent (void) {
		/* YOUR CODE GOES HERE*/
		delete(name);
	}

	/* YOUR CODE GOES HERE */
/***************************************************************************
% Routine Name : operator const char* (void) const
% File :         Driver.c
% 
% Description :  This functin cast a UCSDStudent into a constant char* of its
			name
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
***************************************************************************/
	operator const char * (void) const {
		return name;
	}
/***************************************************************************
% Routine Name : operator long (void) const
% File :         Driver.c
% 
% Description :  This functin cast a UCSDStudent into a long integer of its
			hase code
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
***************************************************************************/
	operator long (void) const{
		/* stores the returning value */
		long sum = 0;
		/* loop through each char in name and add the ascii value to
		 * sum */
		for(long index = 0; name[index]; index ++)
			sum += name[index];
		return sum;
	}
/***************************************************************************
% Routine Name : operator == (const Base & base)const
% File :         Driver.c
% 
% Description :  This functin cast override == operator so it compares the
%			name of two UCSDStudents and return true if equal
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
***************************************************************************/
	long operator == (const Base & base)const {
		return !strcmp (name, base);
	}
/***************************************************************************
% Routine Name : operator < (const Base & base)const
% File :         Driver.c
% 
% Description :  This functin cast override < operator so it compares the
%			name of two UCSDStudents and return true if one is less
%
% Parameters descriptions :
% 
% name              	description
% ------------------ ------------------------------------------------------
***************************************************************************/
	long operator < (const Base & base) const {
		return (strcmp (name, base) < 0) ? 1 : 0;
	}
	ostream & Write (ostream & stream) const {
		return stream << "name:  " << name
			<< "  Studentnum:  " << studentnum;
	}
};

/* You may have one or more functions defined here, too */



int main (int argc, char * const * argv) {
	SymTab ST (5);
	char buffer[128];
	char command;
	long number;
	char option;

	/* initialize debug variable */
	HashTable::Set_Debug(0);

	 /* check command line options for debug display */
	 while ((option = getopt (argc, argv, "x")) != EOF) {

		 switch (option) {
			 case 'x': HashTable::Set_Debug(1);
				 break;
		 	}
	} 

	ST.Write (cout << "Initial Symbol Table:\n");
	while (cin) {
		command = NULL;		// reset command each time in loop
		cout << "Please enter a command:  ((i)nsert, (l)ookup, (w)rite):  ";
		cin >> command;

		switch (command) {
		case 'i':
			cout << "Please enter UCSD student name to insert:  ";
			cin >> buffer;	// formatted input

			cout << "Please enter UCSD student number:  ";
			cin >> number;

			// create student and place in symbol table
			if(!ST.Insert (new UCSDStudent (buffer, number))){

				ST.Write (cout << "\nFinal Symbol Table:\n");
				exit (0);
			}
			break;

		case 'l': {
			const Base * found;	// whether found or not

			cout << "Please enter UCSD student name to lookup:  ";
			cin >> buffer;	// formatted input

			UCSDStudent stu (buffer, 0);
			found = ST.Lookup (& stu);
			
			if (found)
				found->Write (cout << "Student found!!!\n") << "\n";
			else
				cout << "Student " << buffer << " not there!\n";
			}
			break;

		case 'w':
			ST.Write (cout << "The Symbol Table contains:\n");
		}
	}
	ST.Write (cout << "\nFinal Symbol Table:\n");
}

File: CSE12/hw6/Driver.java
/**************************************************************************** 
 
                                                                Zifan Zhang 
                                                         	CSE 12, Fall 19 
                                                                11/5/2019 
                                                             	cs12fa19gt 
                		Assignment SIX
 
File Name:  	Driver.java
Description:    This pogram implement UCSDStudent to store student numbers
		with the associated student name
****************************************************************************/
import java.io.*;
//=========================================================================
// class UCSDStudent extends Base
//
// Description: Implements the UCSDStudent. Each UCSDStudent has a name
// 		and a student number. 
//
// Data Fields:
//     name (String)  - name of the UCSD student
//     studentnum (long)  - the student number of the UCSD student
//
// Public functions:
//     UCSDSudent - allocae memory and initialize the UCSDStudent
//     getName - return the name of the UCSDStudent
//     hashCode - return the hash code of the UCSDStudent
//     equals - return true if two UCSDStudent's name are equal
//     isLessThan - return true if a UCSDStudent's name is less than another's
//     toString - return a string that prints the data fields in UCSDStudent
//==========================================================================
class UCSDStudent extends Base {
        private String name; //name of the UCSDStudent
        private long studentnum; //student number of the UCSDStudent

        /* YOUR CODE GOES HERE */
	/**
         * UCSDStudent
         * Allocates and initializes the memory for a UCSDStudent
         *
         * @param  buffer	The name of the UCSDStudent
	 * @param  number	The student number of the UCSDStudent
         */
	public UCSDStudent (String buffer, long number)
	{
		name = new String (buffer);
		studentnum = number;
	}
	/**
         * getName
         * return the name of UCSDStudent
         *
         * <return>   The name of UCSDStudent
         */
	public String getName()
	{
		return name;
	}
	/**
         * hashCode
         * return the hash code of UCSDStudent
         *
         * <return>   The hash code of UCSDStudent
         */
	public int hashCode () 
	{
                int retval = 0; //stores return value
                int index = 0; //counts number of characters in name added
		//loop through each char in name and add its ascii to retval
                while (index != name.length ()) {
                        retval += name.charAt (index);
                        index ++;
                }
		
                return retval;
        }
	/**
         * equals
         * return true if two UCSDStudents have the same name
         *
         * @param  object	The other UCSDStudent
	 * @<reurn>  	true if their name are the same false otherwise
         */

	public boolean equals (Object object)
	{
		Base bbb = (Base) object;
		return name.equals(bbb.getName());
	}
	/**
         * isLessThan
         * return true if the UCSDStudent's name is less than the another's
         *
         * @param  object	The other UCSDStudent
	 * @<reurn>  	true if name is less than the other UCSDStudent's
         */
	public boolean isLessThan (Base bbb)
	{
		return (name.compareTo (bbb.getName ()) < 0) ? true : false;
	}
	/**
         * toString
         * return the data fields of UCSDStudent
         *
	 * @<reurn>  	data fields of UCSDStudent
         */
	public String toString()
	{
		return "name: " + name + " Studentnum: " + studentnum;
	}
}

public class Driver {
        private static final short NULL = 0;

        public static void main (String [] args) {
                /* initialize debug states */
                HashTable.setDebugOff();

                /* check command line options */
                for (int index = 0; index < args.length; ++index) {
                        if (args[index].equals("-x"))
                                HashTable.setDebugOn();
                }

                /* The real start of the code */
                SymTab symtab = new SymTab (5);
                String buffer = null;
                char command;
                long number = 0;

                System.out.print ("Initial Symbol Table:\n" + symtab);

                while (true) {
                        command = 0;    // reset command each time in loop
                        System.out.print ("Please enter a command:  "
                                + "((i)nsert, (l)ookup, (w)rite):  ");

                        try {
                        command = MyLib.getchar ();
                        MyLib.clrbuf (command); // get rid of return

                        switch (command) {
                        case 'i':
                                System.out.print (
                                "Please enter UCSD Student name to insert:  ");
                                buffer = MyLib.getline ();// formatted input

                                System.out.print (
                                        "Please enter UCSD Student number:  ");

                                number = MyLib.decin ();
                                MyLib.clrbuf (command); // get rid of return

                                // create Student and place in symbol table
                                if(!symtab.insert (
                                        new UCSDStudent (buffer, number))){

                                        System.out.print ("\nFinal Symbol "
                                                        + "Table:\n" + symtab);
                                        System.exit(0);
                                }
                                break;

                        case 'l': {
                                Base found;     // whether found or not

                                System.out.print (
                                "Please enter UCSD Student name to lookup:  ");

                                buffer = MyLib.getline ();// formatted input

                                UCSDStudent stu = new UCSDStudent (buffer, 0);
                                found = symtab.lookup (stu);
                                
                                if (found != null) {
                                        System.out.println ("Student found!!!");
                                        System.out.println (found);
                                }
                                else
                                        System.out.println ("Student " + buffer
                                                + " not there!");
                                }
                                break;

                        case 'w':
                                System.out.print (
                                    "The Symbol Table contains:\n" + symtab);
                        }
                        }
                        catch (EOFException eof) {
                                break;
                        }
                }

                System.out.print ("\nFinal Symbol Table:\n" + symtab);
        }
}
